feat: Create upgrading tool (#988)
vaadin/docs: articles/upgrading/fusion/18-19.asciidoc

[discrete]
==== Fusion Breaking Changes

[discrete]
===== Generated @Id Field Is Now of Optional Type in TypeScript
A field with `@Id` annotation in Java is now of optional type in the generated TypeScript code.
Given an entity with an `id` field:

.`Entity.java`
[source,java]
----
public class Entity {
    @Id
    private int id;
}
----

Now in the TypeScript files, instead of using `endpoint.getEntity(entity.id)`, you might need to change to `endpoint.getEntity(entity.id!)` (if you know that the `id` is always set when this is called) or add a type guard to explicitly check that `id` is not `undefined`.

[discrete]
==== Ignore One More Service Worker Related Static File
You need to ignore one more static file, `/sw-runtime-resources-precache.js`, if you use `HttpSecurity.authorizeRequests()` to do role-based authorization in your security configuration as follows:

.`SecurityConfiguration.java`
[source,java]
----
@Override
protected void configure(HttpSecurity http) throws Exception {
    ...
    http.authorizeRequests().anyRequest().hasAnyAuthority(Role.getAllRoles());
    ...
}
----

> In such case, you need to add one more file `/sw-runtime-resources-precache.js` to the static resource list that Spring Security bypasses:

Reviewer: ```suggestion
In this situation, you need to add one more file `/sw-runtime-resources-precache.js` to the static resource list that Spring Security bypasses:

```
